routing/base/astar_algorithm.hpp

```diff
@@ -114,30 +140,31 @@ class AStarAlgorithm
                      VisitVertex && visitVertex, Context & context) const;
 
   Result FindPath(TGraphType & graph, TVertexType const & startVertex,
-                  TVertexType const & finalVertex, RoutingResult<TVertexType> & result,
+                  TVertexType const & finalVertex, RoutingResult<TVertexType, TWeightType> & result,
                   my::Cancellable const & cancellable = my::Cancellable(),
                   TOnVisitedVertexCallback onVisitedVertexCallback = nullptr) const;
 
   Result FindPathBidirectional(TGraphType & graph,
                                TVertexType const & startVertex, TVertexType const & finalVertex,
-                               RoutingResult<TVertexType> & result,
+                               RoutingResult<TVertexType, TWeightType> & result,
                                my::Cancellable const & cancellable = my::Cancellable(),
                                TOnVisitedVertexCallback onVisitedVertexCallback = nullptr) const;
 
   // Adjust route to the previous one.
   // adjustLimit - distance limit for wave propagation, measured in same units as graph edges length.
   typename AStarAlgorithm<TGraph>::Result AdjustRoute(
       TGraphType & graph, TVertexType const & startVertex, vector<TEdgeType> const & prevRoute,
-      double adjustLimit, RoutingResult<TVertexType> & result, my::Cancellable const & cancellable,
-      TOnVisitedVertexCallback onVisitedVertexCallback) const;
+      TWeightType adjustLimit, RoutingResult<TVertexType, TWeightType> & result,
+      my::Cancellable const & cancellable, TOnVisitedVertexCallback onVisitedVertexCallback) const;
 
 private:
   // Periodicity of switching a wave of bidirectional algorithm.
   static uint32_t constexpr kQueueSwitchPeriod = 128;
 
   // Precision of comparison weights.
-  static double constexpr kEpsilon = 1e-6;
-  static double constexpr kInfiniteDistance = numeric_limits<double>::max();
+  static TWeightType constexpr kEpsilon = GetWeightEpsilon<TWeightType>();
+  static TWeightType constexpr kZeroDistance = GetWeightZero<TWeightType>();
+  static TWeightType constexpr kInfiniteDistance = GetWeightMax<TWeightType>();
```

Comment:
I see two options:
1) In case TWeightType does not have "infinity" it can have flag inside which means "it's infinity" and TWeightType's operator < can respect this flag. So it'll not be AStar problem.
2) Remove kInfiniteDistance and GetWeightMax and rewrite AStarAlgorithm::Context::GetDistance to return pair<TWeightType, bool>. It's really not hacky at all but it leads to make GetDistance calls less clear.

So I just not sure about making existent code less clear to satisfy some hypothetical restriction about TWeightType which does not have "infinity". Or maybe there are the third option which I do not see?